toml/src/main/kotlin/org/rust/toml/intentions/AddCrateDependencyIntention.kt

```diff
@@ -0,0 +1,129 @@
+import org.rust.ide.experiments.RsExperiments
+import org.rust.lang.core.psi.RsExternCrateItem
+import org.rust.lang.core.psi.RsPath
+import org.rust.lang.core.psi.RsUseItem
+import org.rust.lang.core.psi.ext.*
+import org.rust.openapiext.isFeatureEnabled
+import org.rust.openapiext.toPsiFile
+import org.rust.toml.crates.local.CargoRegistryCrate
+import org.rust.toml.crates.local.CratesLocalIndexException
+import org.rust.toml.crates.local.CratesLocalIndexService
+import org.toml.lang.psi.*
+
+class AddCrateDependencyIntention : TomlElementBaseIntentionAction<Context>() {
+    override fun getText(): String = "Add crate to dependencies"
+    override fun getFamilyName(): String = text
+
+    override fun findApplicableContextInternal(project: Project, editor: Editor, element: PsiElement): Context? {
+        if (!isFeatureEnabled(RsExperiments.CRATES_LOCAL_INDEX)) return null
+
+        val externCrateItem = element.parentOfType<RsExternCrateItem>(true)
+        val path = element.parentOfType<RsPath>(true)
+
+        val target: RsElement = when {
+            externCrateItem != null -> {
+                if (externCrateItem.reference.multiResolve().isNotEmpty()) return null
+                externCrateItem
+            }
+            path != null -> {
+                if (path.reference == null) return null
+                if (path.parentOfType<RsUseItem>() == null) return null
+
+                val isPathUnresolved = path.resolveStatus != PathResolveStatus.RESOLVED
+                if (!isPathUnresolved) return null
+
+                val qualifier = path.qualifier
+                if (qualifier != null) return null
+                path
+            }
+            else -> null
+        } ?: return null
+
+        if (target.containingCrate?.origin != PackageOrigin.WORKSPACE) return null
+
+        val (crateName, crate) = getAvailableCrate(target) ?: return null
+        return Context(target, crateName, crate)
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val element = ctx.element
+        val pkg = element.containingCargoPackage ?: return
+        val manifest = pkg.contentRoot?.findChild("Cargo.toml") ?: return
+        val tomlManifest = manifest.toPsiFile(project) as? TomlFile ?: return
+
+        val version = ctx.crate.sortedVersions.reversed().firstOrNull { !it.isYanked } ?: return
+
+        val factory = TomlPsiFactory(project)
+        val dependencies = findOrCreateDependencies(tomlManifest, factory)
+
+        val dependency = factory.createKeyValue(ctx.crateName, "\"${version.version}\"")
+
+        val lastChild = dependencies.entries.lastOrNull()
+        val inserted = if (lastChild != null) {
+            dependencies.addAfter(dependency, lastChild)
+        } else {
+            dependencies.add(dependency)
+        }
+        dependencies.addBefore(factory.createNewline(), inserted)
+
+        tomlManifest.navigate(true)
```

Comment:
Is it the desired behaviour that caret/editor doesn't focus on added dependency?